src/charm_state.py

```diff
@@ -0,0 +1,141 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The module for managing charm state."""
+
+
+import logging
+import os
+import typing
+from dataclasses import dataclass
+
+import ops
+from pydantic import BaseModel, ValidationError
+
+import metadata
+
+# agent relation name
+AGENT_RELATION = "agent"
+
+logger = logging.getLogger()
+
+
+class Credentials(BaseModel):
+    """The credentials used to register to the Jenkins server.
+
+    Attrs:
+        address: The Jenkins server address to register to.
+        secret: The secret used to register agent.
+    """
+
+    address: str
+    secret: str
+
+
+class CharmStateBaseError(Exception):
+    """Represents error with charm state."""
+
+
+class InvalidStateError(CharmStateBaseError):
+    """Exception raised when state configuration is invalid."""
+
+    def __init__(self, msg: str = ""):
+        """Initialize a new instance of the InvalidStateError exception.
+
+        Args:
+            msg: Explanation of the error.
+        """
+        self.msg = msg
+
+
+def _get_jenkins_unit(
+    all_units: typing.Set[ops.Unit], current_app_name: str
+) -> typing.Optional[ops.Unit]:
+    """Get the Jenkins charm unit in a relation.
+
+    Args:
+        all_units: All units in a relation.
+        current_app_name: The Jenkins-agent application name.
+
+    Returns:
+        The Jenkins server application unit in the relation if found. None otherwise.
+    """
+    for unit in all_units:
+        # if the unit's application name is the same, this is peer unit. Otherwise, it is the
+        # Jenkins server unit.
+        if unit.app.name == current_app_name:
+            continue
+        return unit
+    return None
```

Comment:
Similar to below, isn't this an error? That we got relation data but it didn't include data for this unit?

Reply:
This can happen in cases where there are multiple units of the charm, I also mentioned it in details here: https://github.com/canonical/jenkins-agent-operator/pull/4#discussion_r1427750214